batcher.go

```diff
@@ -0,0 +1,327 @@
+			wl.Error(ErrShutdown)
+			wl.Close()
+		}
+		return
+	}
+
+	b.pending = append(b.pending, waiterlist...)
+
+	b.runBatch(MainBatcher)
+}
+
+// func (b *Batcher[K, V]) load(ctx context.Context, key K, result chan *Result[K, V]) {
+// 	b.mu.Lock()
+// 	defer b.mu.Unlock()
+
+// 	if b.shutdown {
+// 		result <- &Result[K, V]{Key: key, Err: ErrShotdown}
+// 	}
+
+// 	b.pending = append(b.pending, waiter[K, V]{ctx: ctx, key: key, res: result})
+
+// 	b.runBatch(MainBatcher)
+// }
+
+// runBatch checks the number of valid threads and pending data and starts a new thread to process the data.
+// b.mu must be lock before
+func (b *Batcher[K, V]) runBatch(btype batcherType) {
+	// If the number of threads is the maximum, we do not create a new one.
+	// We do not create a batcher "TimerBatcher", because the maximum number of threads shows
+	// the optimal operation of the system (all keys are taken)
+	if b.batcherCount >= b.opt.MaxBatcher {
+		return
+	}
+
+	batch := b.nextBatch(btype)
+	if len(batch) > 0 {
+		b.wg.Add(1)
+		go func() {
+			b.callHandlers(btype, batch)
+			b.wg.Done()
+		}()
+		b.batcherCount++
+	}
+}
+
+// nextBatch fetches data from pending.
+// b.mu must be lock before
+func (b *Batcher[K, V]) nextBatch(btype batcherType) []*waiter[K, V] {
+	if len(b.pending) < b.opt.MinBatchSize && btype != TimerBatcher {
+		return nil
+	}
+
+	// Send all
+	if b.opt.MaxBatchSize == 0 || len(b.pending) <= b.opt.MaxBatchSize {
+		batch := b.pending
+		b.pending = nil
+
+		return batch
+	}
+
+	batch := make([]*waiter[K, V], 0, len(b.pending))
+	for _, msg := range b.pending {
+		if b.opt.MaxBatchSize > 0 && len(batch)+1 > b.opt.MaxBatchSize {
+			break
+		}
+
+		batch = append(batch, msg)
+	}
+	b.pending = b.pending[len(batch):]
+
+	return batch
+}
+
+// callHandlers prepare context, slice of keys and run user handler.
+// After processing the batch tries to get a new data packet. If pending data is empty
+func (b *Batcher[K, V]) callHandlers(btype batcherType, batch []*waiter[K, V]) {
+	for len(batch) > 0 {
+		ctx := batch[0].ctx
+
+		keys := make([]K, 0, len(batch))
+		for _, item := range batch {
+			keys = append(keys, item.key)
+		}
+
+		items := b.handler(ctx, keys)
+		if len(keys) != len(items) {
+			err := fmt.Errorf(`
+				The batch function supplied did not return an array of responses
+				the same length as the array of keys.
+
+				Keys:
+				%v
+
+				Values:
+				%v
+			`, keys, items)
+
+			for _, wt := range batch {
+				wt.Error(err)
+			}
+		} else {
+			for i := range batch {
+				batch[i].Response(items[i])
+			}
+		}
+
+		b.mu.Lock()
+
+		// At the first start TimerBatcher takes all the data from pending, the size of which is less than the minimum size batch size.
+		// In order for this batcher to work according to the general rules in the future, we change its type.
+		if btype == TimerBatcher {
+			btype = MainBatcher
+		}
+		batch = b.nextBatch(btype)
+		if len(batch) == 0 {
+			b.batcherCount--
```

Comment:
I would move it to `runBatch` inside `go func` just for ++ and -- to be as close as possible and not to overcomplicate the `callHandlers`. WDYT?